### PR TITLE
Add [BOXKeychainItemWrapper allKeychainItemWrappersWithAccessGroup:] method.

### PR DESCRIPTION
**Pull request by RicoYao**
This retrieves an array of BOXKeychainItemWrapper objects from the keychain.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @RicoYao has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/External/KeychainItemWrapper/BOXKeychainItemWrapper.m [hunk 1]
```diff
@@ -400,4 +395,41 @@ - (NSMutableDictionary *)defaultKeychainItemDataDictionaryWithIdentifier:(NSStri
     return defaultDictionary;
 }
 
+// Retrieve all keychain entries that have been made by instances of BOXKeychainItemWrapper.
+// Sample usage in arc land: NSArray* keychainEntries = (__bridge_transfer NSArray*)[BOXKeychainItemWrapper allEntries];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/External/KeychainItemWrapper/BOXKeychainItemWrapper.m]: allEntries -> allKeychainItemWrappersWithAccessGroup:acccessGroup

File: BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m [hunk 2]
```diff
@@ -204,56 +204,38 @@ + (NSArray *)usersInKeychain
...
-    {
-        NSArray* keychainEntries = (__bridge_transfer NSArray*)keychainQueryResult;
-        for (NSDictionary *dict in keychainEntries)
-        {
-            NSObject *object = [dict objectForKey:((__bridge id)kSecAttrGeneric)];
-            if (![object isKindOfClass:[NSString class]]) {
-                continue;
-            }
+    // Go through all keychain items added by BOXKeychainItemWrapper
+    NSArray *keychainItemWrappers = [BOXKeychainItemWrapper allKeychainItemWrappersWithAccessGroup:[self keychainAccessGroup]];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m:208]: (__bridge_transfer NSArray *)
?

File: BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m [hunk 3]
```diff
@@ -204,56 +204,38 @@ + (NSArray *)usersInKeychain
...
-            {
-                NSString *userID = [self userIDFromKeychainIdentifier:keychainIdentifier];
-                if (userID.length > 0)
-                {
-                    NSString *jsonString = [[self keychainItemWrapperForUserWithID:userID] objectForKey:(__bridge id)kSecValueData];
+            // Only look at those that have our prefix
+            if ([keychainIdentifier hasPrefix:[self keychainIdentifierPrefix]]) {
+                
+                // Extract user-id from prefix, and credentials from keychain value data
+                NSString *userID = [self userIDFromKeychainIdentifier:(NSString *)keychainIdentifier];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m]: Don't need the NSString cast here.

File: BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m [hunk 4]
```diff
@@ -204,56 +204,38 @@ + (NSArray *)usersInKeychain
...
+                        NSString *userIDFromKeychain = [dictionary objectForKey:keychainUserIDKey];
+                        NSString *userNameFromKeychain = [dictionary objectForKey:keychainUserNameKey];
+                        NSString *userLoginFromKeychain = [dictionary objectForKey:keychainUserLoginKey];
+                        if ([userID isEqualToString:userIDFromKeychain]) {
+                            BOXUserMini *miniUser = [[BOXUserMini alloc] initWithUserID:userIDFromKeychain
+                                                                                   name:userNameFromKeychain
+                                                                                  login:userLoginFromKeychain];
+                            [users addObject:miniUser];
+                        }
                     }
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/OAuth2/BOXAbstractSession.m:233]: Anything we should do in the case of an error? returning nil ultimately for 'users' could be potentially misleading if there are indeed entries in the keychain...

**Comment by boxtcarpel** [comment]
+2 LGTM
:bookmark_tabs: 

Great improvement!

File: BoxContentSDK/BoxContentSDK/External/KeychainItemWrapper/BOXKeychainItemWrapper.m (hunk 1 above)
Comment (RicoYao) [inline BoxContentSDK/BoxContentSDK/External/KeychainItemWrapper/BOXKeychainItemWrapper.m]: Whoops, this was a comment for an older version of this method. I added the correct comment to the .h. Will remove this.

File: BoxContentSDK/BoxContentSDK/External/KeychainItemWrapper/BOXKeychainItemWrapper.m [hunk 5]
```diff
@@ -400,4 +395,39 @@ - (NSMutableDictionary *)defaultKeychainItemDataDictionaryWithIdentifier:(NSStri
...
+    if (accessGroup.length > 0) {
+        [keychainQuery setObject:accessGroup forKey:(__bridge id)kSecAttrAccessGroup];
+    }
+#endif
+    
+    NSMutableArray *keychainItemWrappers = [NSMutableArray array];
+    
+    CFArrayRef keychainQueryResult = NULL;
+    OSStatus queryStatus =  SecItemCopyMatching((__bridge CFDictionaryRef)keychainQuery, (CFTypeRef *)&keychainQueryResult);
+    if (queryStatus == 0 && keychainQueryResult != NULL) {
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/External/KeychainItemWrapper/BOXKeychainItemWrapper.m:418]: Use 'errSecSuccess' ?

**Comment by boxtcarpel** [comment]
+2 LGTM

:wine_glass:

**Comment by RicoYao** [comment]
Likely too stale now (though still looks right conceptually). Closing.

**Pull request closed without merging**